### DISCUSSION TITLE
Align (rotated) objects inside a group - issue

### DISCUSSION THREAD

**Discussion by val3n71na** [fabricjs/fabric.js#9721] [Q&A] [answered]
Hello,

I am trying to align objects inside a group to left/right and it works well until the objects are **ROTATED**. I've noticed that width and height are remaining the same, even if the object is rotated and visually width and height are not as they were at the beginning.

You can see the problem here: https://jsfiddle.net/IoanaBraica/56b7qp1c/41/
1. Select both objects (circle & triangle)
2. Click align left -> ALL GOOD, the alignment works
3. Deselect group. Select again the 2 objects
4. Click align right -> ALL GOOD, the alignment works
5. Deselect active selection. ROTATE the triangle
6. Repeat (steps 1 & 2) -> Rotation does not work anymore as expected - same for alignment to the right.

How can I implement the alignment inside a group to also work when some of the objects are rotated? 

Thank you!! :)
![image](https://github.com/fabricjs/fabric.js/assets/26135080/0b2e97e3-3d84-409b-bfce-6975d1886a93)

**Comment by asturur** [comment]
For me the simple solution is that you take all the objects you are interested in and you calculate the bounding rect of each.

Object.getBoundingRect()

Once you have the bounding rect that is an object with { left , top , width, height }, you take the smallest left of all, that is your alignment point.

Once you have that for each object you move the left property of the difference between the smallest left and the left of it bounding rect.

```
const boxA = a.getBoundingRect();
const boxB = b.getBoundingRect();

const minLeft = Math.min(boxA.left, boxB.left);

a.left -= boxA.left - minLeft;
b.left -= boxB.left - minLeft;

```

Something similar to what i did here should allow you to build a solution

I see that in fabric v6 the bounding box is always in scene plane, so if the group has a transformation this won't work at all.
In case you have to build the bounding box with the aCoords object:

```
const getRelativeBoundingRect = (object) =>  return makeBoundingBoxFromPoints(Object.values(object.aCoords));
```

**Reply by val3n71na** [reply]
Thank you for your response! 

Using the boundingRect (fabric 5.3.0) I get the following issue - the alignment is not as expected:
![image](https://github.com/fabricjs/fabric.js/assets/26135080/10822e47-e0de-419d-8287-15cc2f30a942)

I am going to upgrade to v6 and check your solution building the bounding box with the aCoords object. I will be back with an update.

**Reply by asturur** [reply]
not sure what you solution want is exactly what i thought you wanted to achieve and i gave you code for.
Is the issue because the triangle has a vertex in the middle of the box and so to you the triangle box seems unnecessary larger?
With v6 you are getting the same exact result. if you want to align items by their own vertext, that will be extremely different and complicated.

**Reply by val3n71na** [reply]
Hello, yes, that is exactly what I need. I need the triangle (and more other irregular shapes) to be aligned by its own vertex instead of its bounding box.

**Reply by asturur** [reply]
You can solve it for polygons/polylines and you can do something clever for circles with the example below from @ShaMan123 but there is no utility ( and probably won't be in this library ) to do the same with random paths.
On top of that, even if you solve the calculation issue, you have to take in account that the group selection will adapt itself to content but it will take in account the larger bounding box.

The general concept is:

use object.calcTransformMatrix() to find the current transform of the object
transforms the points of the polygon to a common space ( the parent space usually if you want to align inside a generic group or selection)
get the bounding rect of every group of vertex
find the alignment point
shift object by the difference with the alignment point

**Comment by ShaMan123** [comment, marked as answer]
In order to get the vertex in the scene plane do the following
```typescript
const { pathOffset: { x, y } } = poly;
const polyVertexTransform = multiplyTransformMatrices(poly.calcTransformMatrix(), createTranslateMatrix(-x, -y))
const polyVertices = poly.points.map(p => sendPointToPlane(new Point(p), polyVertexTransform))
const polyBBox = makeBoundingBoxFromPoints(polyVertices)
```

**Reply by ShaMan123** [reply]
This can work with Path as well but needs bezier-js, see #9511

**Reply by ShaMan123** [reply]
don't forget about stroke projection though, bezier-js can handle that for you with ease

**Reply by val3n71na** [reply]
Thank you! I am going to try these solutions. 👍
